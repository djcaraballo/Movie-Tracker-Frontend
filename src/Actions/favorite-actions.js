export const addFavorite = (movie, user) => ({
  type: 'ADD_FAVORITE',
  movie_id: movie.id,
  user_id: user.id,
  title: movie.title,
  poster_path: movie.poster_path,
  release_date: movie.release_date,
  vote_average: movie.vote_average,
  overview: movie.overview,
});