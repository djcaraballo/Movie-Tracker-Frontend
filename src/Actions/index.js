export const addMovies = (movieInfo) => ({
  type: 'ADD_MOVIES',
  movies: movieInfo
})