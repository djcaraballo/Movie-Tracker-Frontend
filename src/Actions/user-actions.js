export const newUser = (name, email, password) => ({
  type: 'NEW_USER',
  name,
  email,
  password,
});

export const fetchUser = (status) => ({
  type: 'FETCH_USER',
  status,
});

export const toggleUserStatus = (status) => ({
  type: 'TOGGLE_STATUS',
  status: !status,
});
