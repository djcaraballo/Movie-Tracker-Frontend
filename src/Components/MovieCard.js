import React from 'react';

const MovieCard = (props) => {

  return (
    <div className='movie-card'>
      <img src={props.poster_path} alt='movie poster' />
      <h1>{props.title}</h1>
      <p>{props.release_date}</p>
      <p>{props.overview}</p>
      <p>{props.vote_average}</p>
    </div>
  )
}

export default MovieCard;