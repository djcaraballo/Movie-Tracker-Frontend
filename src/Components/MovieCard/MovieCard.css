.movie-poster:hover {
  box-shadow: #eaffea 0px 0px 20px 5px;
  cursor: pointer;
}

.movie-title {
  color: #bdcdcd;
  font-family: 'Lato', sans-serif;
  font-size: 1.5em;
  font-weight: 400;
}