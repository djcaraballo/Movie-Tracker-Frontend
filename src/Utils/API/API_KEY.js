export const key = '5cb4969ffac5d2d967beb6c31aa53509';