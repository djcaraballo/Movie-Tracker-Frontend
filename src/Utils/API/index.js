import { key } from './API_KEY';

export const fetchMovies = async() => {
  const url = `https://api.themoviedb.org/3/movie/now_playing?api_key=${key}&language=en-US&page=1`;
  const response = await fetch(url);
  const result = await response.json();
  return result;
}

export const postNewUser = async (newUserInfo) => {
  fetch('http://localhost:3000/api/users/new',{ 
    method: "POST",
    body: JSON.stringify({...newUserInfo, favorites: []}),
    headers:{'Content-Type': 'application/json'},
  })
  .then(res => res.json())
  .then(response => console.log('Success:', JSON.stringify(response)))
  .catch(error => console.error('Error:', error));
}

export const signInUser= async (userInfo) => {
  fetch('http://localhost:3000/api/users',{
	method: "POST",
	body: JSON.stringify(userInfo),
	headers:{'Content-Type': 'application/json'},
  })
  .then(res => res.json())
  .then(response => console.log('Success:', JSON.stringify(response)))
  .catch(error => console.error('Error:', error));
}
