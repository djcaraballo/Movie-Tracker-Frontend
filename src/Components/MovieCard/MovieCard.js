import React, {Component} from 'react';
import {addFavorite} from '../../Actions/favorite-actions';
import {connect} from 'react-redux';
import * as API from '../../Utils/API/';
import hollowStar from '../../Images/star.svg';
import solidStar from '../../Images/bookmark-star.svg';
import './MovieCard.css';


class MovieCard extends Component {
  constructor(props) {
    super();
    this.state = {
      favorite: false,
      userFavorites: [],
    };
  }

  async componentDidMount() {
    console.log(this.props.user.id);
    let favoriteList = await API.getFavorites(this.props.user.id);
    this.setState({userFavorites: favoriteList}, () => this.handleFavorites());
  }

  handleFavorites = () => {
    const {userFavorites} = this.state;
    userFavorites.map(fav => {
      if (fav.title === this.props.title) {
        this.setState({favorite: true});
      } else {
        return;
      }
    });
  };

  handleClick = () => {
    const {
      title,
      poster_path,
      id,
      release_date,
      vote_average,
      overview,
    } = this.props;
    this.setState({favorite: !this.state.favorite});
    this.props.handleFavorite(
      {title, poster_path, id, release_date, vote_average, overview},
      this.props.user,
    );
    let favoriteMovie = {
      title,
      poster_path,
      user_id: this.props.user.id,
      movie_id: id,
      release_date,
      vote_average,
      overview,
    };
    if (!this.state.favorite) {
      API.addFav(favoriteMovie);
    } else if (this.state.favorite) {
      API.removeFavorites(favoriteMovie);
    };
  };

  render() {
    const {favorite} = this.state;
    if (!favorite) {
      return (
        <div className="movie-card">
          <img
            className="star-icon"
            onClick={this.handleClick}
            src={hollowStar}
            alt="Favorite not selected"
          />
          <img src={this.props.poster_path} alt="movie poster" />
          <h1>{this.props.title}</h1>
        </div>
      );
    } else if (favorite) {
      return (
        <div className="movie-card">
          <img
            className="fav-star-icon"
            onClick={this.handleClick}
            src={solidStar}
            alt="Favorite not selected"
          />
          <img src={this.props.poster_path} alt="movie poster" />
          <h1>{this.props.title}</h1>
        </div>
      );
    }
  }
}

const mapStateToProps = state => {
  return {
    favorites: state.favorites,
    user: state.user,
  };
};

const mapDispatchToProps = dispatch => ({
  handleFavorite: (movie, user) => {
    dispatch(addFavorite(movie, user));
  },
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(MovieCard);
