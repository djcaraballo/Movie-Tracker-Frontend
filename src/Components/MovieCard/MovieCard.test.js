import React from 'react';
import ReactDOM from 'react-dom';
import {connect} from 'react-redux';
import * as API from '../../Utils/API'
import {MovieCard, mapStateToProps, mapDispatchToProps} from './MovieCard';
import { updateMovies } from '../../Actions';
import { shallow } from 'enzyme';

describe('MovieCard', () => {
  let wrapper
  let mockMovie = { id: 7, poster_path: '/something.jpg', title: 'MOVIE', favorite: false }
  const store = jest.fn()
  
  beforeEach(() => {
    wrapper = shallow(<MovieCard key={1} {...mockMovie} />)
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  });

  describe('mapFavorites', () => {
    it('should fire toggleFavoriteInState if title matches', () => {
      wrapper.instance().toggleFavoriteInState = jest.fn()
      wrapper.instance().mapFavorites([mockMovie])
    })
  });

  describe('handleClick function', () => {
    it('should switch state to true favorited', () => {
      wrapper.instance().toggleFavoriteInState = jest.fn()
      expect(wrapper.state().favorite).toEqual(false)
      wrapper.instance().mapFavorites([mockMovie])
      expect(wrapper.state().favorite).toEqual(true)
    });
  });
  
  describe('handleFavorite function', () => {
    it('should fire addFav if favorite in state is false', () => {
      API.addFav = jest.fn()
      let mockUser = { id: 7 }
      let newWrapper = shallow(<MovieCard key={1} {...mockMovie} user={mockUser} />)
      newWrapper.instance().handleFavorite(mockMovie)
      expect(API.addFav).toHaveBeenCalled()
    });
  
    it('should fire removeFavorite if favorite in state is true', () => {
      API.removeFavorite = jest.fn()
      let mockUser = { id: 7 }
      let newWrapper = shallow(<MovieCard key={1} {...mockMovie} user={mockUser} />)
      newWrapper.instance().toggleFavoriteInState = jest.fn()
      newWrapper.instance().setState({ favorite: true })
      newWrapper.instance().handleFavorite(mockMovie)
      expect(API.removeFavorite).toHaveBeenCalled()
    });
  });

  describe('mapStateToProps function', () => {
    it('should return an object with the user and movies', () => {
      const mockState = {
        movies: [ { title: 'MOVIE' } ],
        user: {name: 'Al-awesome', email: 'blah@gmail.com'},
        favorites: []
      }
      const expected = {
        user: {name: 'Al-awesome', email: 'blah@gmail.com'},
        movies: [ { title: 'MOVIE' } ]
      }
      const mappedProps = mapStateToProps(mockState)
      expect(mappedProps).toEqual(expected)
    });
  });

  describe('mapDispatchToProps function', () => {
    it('calls dispatch with a updateMovies action when updateMovies is called', () => {
      const mockDispatch = jest.fn()
      const mockDispatchAction = updateMovies([mockMovie])
      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.updateMovies([mockMovie])
      expect(mockDispatch).toHaveBeenCalledWith(mockDispatchAction)
    })
  })
});