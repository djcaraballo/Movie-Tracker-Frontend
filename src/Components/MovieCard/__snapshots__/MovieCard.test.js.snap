// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieCard should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MovieCard
    favorite={false}
    id={7}
    poster_path="/something.jpg"
    title="MOVIE"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <img
          alt="Favorite not selected"
          className="star-icon"
          onClick={[Function]}
          src="star.svg"
        />,
        <img
          alt="movie poster"
          src="/something.jpg"
        />,
        <h1>
          MOVIE
        </h1>,
      ],
      "className": "movie-card",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "alt": "Favorite not selected",
          "className": "star-icon",
          "onClick": [Function],
          "src": "star.svg",
        },
        "ref": null,
        "rendered": null,
        "type": "img",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "alt": "movie poster",
          "src": "/something.jpg",
        },
        "ref": null,
        "rendered": null,
        "type": "img",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "MOVIE",
        },
        "ref": null,
        "rendered": "MOVIE",
        "type": "h1",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <img
            alt="Favorite not selected"
            className="star-icon"
            onClick={[Function]}
            src="star.svg"
          />,
          <img
            alt="movie poster"
            src="/something.jpg"
          />,
          <h1>
            MOVIE
          </h1>,
        ],
        "className": "movie-card",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "alt": "Favorite not selected",
            "className": "star-icon",
            "onClick": [Function],
            "src": "star.svg",
          },
          "ref": null,
          "rendered": null,
          "type": "img",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "alt": "movie poster",
            "src": "/something.jpg",
          },
          "ref": null,
          "rendered": null,
          "type": "img",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "MOVIE",
          },
          "ref": null,
          "rendered": "MOVIE",
          "type": "h1",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
